### ISSUE TITLE
Proposal: Maybe filter tiddlers shouldn't be text/vnd.tiddlywiki, but instead be its own type.

### ISSUE THREAD

**Issue by flibbles** [TiddlyWiki/TiddlyWiki5#4395]
There are a few cases in Tiddlywiki core where tiddlers store multiline filters, like _$:/DefaultTiddlers_, _$:/config/SyncFilter_, and _$:/config/FileSystemPaths_. They are not wikitext, and when they're rendered as wikitext, they render deceptively wrong, and usually filled with bad links. I realize they could just be specified as `text/plain`, but what if it had its own type?

@pmario pointed out to me that the types are correlated to official MIME types, but unofficial types are allowed with the **x-** prefix, like `text/x-tiddler-dictionary`, so I don't think it'd be completely crazy to have a `text/x-tiddler-filter` type.

It seems at first that the gain would be little, but it'd be useful for plugin developers. My [Relink](https://github.com/flibbles/tw5-relink) plugin is messing up with tiddlers like _$:/DefaultTiddlers_, because it has no reason to think it's not wikitext. I _could_ just hardcode an exception, but then there are people like @pmario, who has filter tiddlers in his project, and has been having to work with me to find a way to make my plugin not mess his up. It seems like our solution right now is for him to put his filters in a _filter_ field, but then he's losing the luxury of having multiline tiddlers, which is especially relevant for what he's doing (making filters for preconfigured stories, I think).

Also, I know plugins like TiddlyMap cram multiline filters into a filter field, and then use funky javascript to make it multiline again. It might be nice if it had a nicer way of doing that.

Also, it might be nice if Vanilla Tiddlywiki could render filters nicely. It might just show them plaintext, but if the `text/x-tiddler-filter` existed, it could easily render them more nicely, or have a plugin do so with ease.

What do y'all think?

**Comment by AnthonyMuscio** [comment]
You can put multi-line text into fields, the only problem is the edit template for fields is not aware of this possibility and if you edit it there you will destroy the line feeds. 

For those fields wanting to display multiline content you can include <br> in its content.

Without suggesting your proposal should not also be followed, allowing for a new class of field that permits multi-line text to be maintained, or in fact secretly enabling fields to be text areas rather than single lines may be more beneficial.

You can permit multiline fields but hiding the field from the editor template so it is not used and providing alternate edit methods is needed.

Regards
Tony

**Comment by flibbles** [comment]
Were you trying to say `<br>`? Because that's the only way I found to include newlines in fields, and that's only after it's been parsed as wikitext. And `<br>` breaks any filters if it's included in them.

**Comment by AnthonyMuscio** [comment]
If you edit a field with the edit text with tag=text area invisible line feeds exist. Depending on what you want there are ways to do searches etc for <br>

**Comment by pmario** [comment]
I'm working with a "filter" field now

**Comment by Jermolene** [comment]
Hi @flibbles I think it's not a bad idea to add a content-type for filters, not least because it would avoid them getting wrongly wikified when they are displayed. `text/x-tiddler-filter` makes sense.

@AnthonyMuscio I think your suggestion of allowing for multiline fields is solving a subtly different problem.

**Comment by AnthonyMuscio** [comment]
@Jermolene agreed. Such fields could be hidden from the current field editor and presented in a different one. I just don't know how to flag them as text area fields. But I have an idea.

**Comment by pmario** [comment]
@AnthonyMuscio ... I'm on the way to finish the list-like, filter-like field editor supporting drag & drop for both elements. ... It's part of the "story" workflow.

I think there is no need anymore to create "special tiddlers". ... I think you'll like it.

**Comment by pmario** [comment]
I did just stumbled upon this issue again and did modify my last comment. 

>Also, it might be nice if Vanilla Tiddlywiki could render filters nicely. It might just show them plaintext, but if the text/x-tiddler-filter existed, it could easily render them more nicely, or have a plugin do so with ease.

The problem with experimental mime types as: `text/x-type` is that they are strongly discouraged by IANA (the spec body)
see: https://datatracker.ietf.org/doc/html/rfc6838#section-3.4 paragraph 2

>   However, with the simplified registration procedures described above
   for vendor and personal trees, it should rarely, if ever, be
   necessary to use unregistered types.  Therefore, use of types in the
   "x." tree is strongly discouraged.

So IMO the Vendor tree: https://datatracker.ietf.org/doc/html/rfc6838#section-3.2 should be the preferred option to be used. 

@Jermolene ... Did you every think about to register `text/vnd.tiddlywiki` and possible "parameters" or "subtypes"?

**Comment by Jermolene** [comment]
> @Jermolene ... Did you every think about to register text/vnd.tiddlywiki and possible "parameters" or "subtypes"?

No. Perhaps wrongly, I thought there was no registry for vnd. content types.

The OP is still a great idea, we should do it.

**Comment by AnthonyMuscio** [comment]
There was some discussion in the past of allowing separate fields to also be given a "type" even "mime type". 

It may be of value to check how that progressed so we "do not unnecessarily limit the possible futures".

**Comment by flibbles** [comment]
Mime types for tiddler bodies already exists, and there already exists the means of rendering them differently. I think expanding this idea to tiddlywiki-specific types wouldn't be much of a stretch.

However, I believe that extending this idea to fields would be a step too far. Fields already have types to a limited extent. The mechanism exists. It's why `colour` and `tags` behave differently, and maybe that can be elaborated a little.

But Mime types for fields? This is degenerate lunacy. :)

Fields are (or should be) consistent with their type. So all `list` fields are lists. All `description` fields would be plaintext maybe. And whatever template renders them will know the type and can render them appropriately. The `text` body doesn't get this benefit.

**Comment by AnthonyMuscio** [comment]
> This is degenerate lunacy. :)

I think you need to review the discussions on this before making this statement. If it helps put the idea into perspective;
* This was for multi-line fields that resemble the text field
* This was an optional setting of a mime type for the field, just as the text field has the type field no more, no less.
* If this were a rule, importing such tiddlers the different fields would just work.
* Tiddlers could include their own stylesheet, and other packaged code.
* If you did not know anything about it, you would not know it exists.

**Comment by flibbles** [comment]
I'm just being facetious. I'm sure there are many good arguments for why to implement it. I would still vote no to such a feature though. I would also have voted no to multi-line fields too, but those happened.

**Comment by AnthonyMuscio** [comment]
> I would also have voted no to multi-line fields

I tried to send a private message but have not worked that out. Here is an example of a great use of multi-line fields, but not with mime types. Unzip and Drop this package on tiddlywiki.com [local-viewtemplate.zip](https://github.com/Jermolene/TiddlyWiki5/files/9246873/local-viewtemplate.zip)
*  then open any tiddler.
* add a new field start typing local and select the fieldname local-viewtemplate > add
* You will see a new multi-line field appear in the tiddler edit mode.
* Now you can edit this multi-line field and it is applied as a viewTemplate only on the current tiddler.

**Comment by kookma** [comment]
With due respect, I also agree to keep the current design! A multi line text field and single line other field.
I believe we should keep the design simple and easy to learn and use by non-technical users.

I think advanced users can add features for their specific use to have multi line field.

**Comment by pmario** [comment]
>I think advanced users can add features for their specific use to have multi line field.

Multi-line fields are possible since v5.2.0. We only don't have a default edit-template to use them properly. ... I did make field-editor and the field-visibility plugin, to use them. 

The main problem fro multi-line fields is that the Node.js server saves them as JSON files instead of .tid, since we don't have a format to create multi-line fields in .tid format.

But we are way OT already.

**Comment by AnthonyMuscio** [comment]
> A multi line edit area for text field and single line input for other fields.

My point is do not do anything that stops such possibilities and others have asked for the ability to extend fields. People like myself already have solutions that do this so backwards compatibility is needed. So if you are going to respond to the OT do so in a way that expands the possibilities not diminish them.

**Comment by flibbles** [comment]
I would once again like to bump this.

I just coded up a "*Relink* Missing" SideBar because I thought it'd be useful, but it's not, largely because it fills up with garbage from all the core filter tiddlers getting parsed as wikitext because there is no way to tell them apart from filters. Take `$:/config/ViewTemplateBodyFilters/system`:

```
[prefix[$:/boot/]] [prefix[$:/config/]] [prefix[$:/core/]!field:title[$:/core/readme]!field:title[$:/core/icon]]
[prefix[$:/info/]] [prefix[$:/language/]] [prefix[$:/languages/]] [prefix[$:/snippets/]] [prefix[$:/state/]]
[prefix[$:/status/]] [prefix[$:/info/]] [prefix[$:/temp/]] +[limit[1]then[$:/core/ui/ViewTemplate/body/code]]
-[prefix[$:/config/Hypothetical]]
```

Every single one of those "$:/..." bits is getting interpreted as a SysLink. Looking at this tiddler directly, and you see it speckled with tc-tiddlylink blue. Deal with it as a Relinking plugin, and you've got a whole lot of missing tiddlers, which you might rename if they ever existed by accident.

It doesn't have to be `text/x-tiddler-filter`, but as long as there's **some** way to differentiate filter tiddlers.

**Comment by pmario** [comment]
https://tiddlywiki.com/#%24%3A%2Fconfig%2FViewTemplateBodyFilters%2Fsystem ... for me it looks like plain text.

**Comment by pmario** [comment]
IMO the OP is related to: https://github.com/Jermolene/TiddlyWiki5/issues/345#issuecomment-33601108

So we would need to implement a type eg: `text/vnd.tiddlywiki; vocab=tiddler-name-or-url;` where the 

`vocab` tiddler contains  pragmas eg: `\rules` that tell the wiki how to parse the text content 

@Jermolene ... We would need to create a new type-parser that is able to parse the name/value syntax and silently ignore everything that it doesn't know. 

@flibbles IMO the type for a tiddler you suggest may be `text/vnd.tiddlywiki; vocab=test`

where test contains

```
\rules only filter ... or filter-visualize
```

@Jermolene ... we would need a new parser rule, that would identify filter strings and which would allow us to show it in a custom format. ... eg: `\rule filter-visualize` ... which may show it in debug mode, discussed at:  **[IDEA] TiddlyWiki filter visualizer/helper tool** #5058

**Comment by pmario** [comment]
@Jermolene ... The transclusion widget may need to respect this new type parameter. 

I'm not use if the widget need a new parameter to overwrite / disable the vocab parameter

**Comment by Jermolene** [comment]
Hi @pmario I don't think that https://github.com/Jermolene/TiddlyWiki5/issues/345#issuecomment-33601108 is relevant here. The problem we're solving #345 is identifying and rendering different variants of wikitext. Here we're considering a completely new content type for filters. Filters are not wikitext, and so they should not share the same `text/vnd.tiddlywiki` type, but instead should get their own type.

Perhaps the argument for reusing `text/vnd.tiddlywiki` is that it avoids us having to mint another content type, but the core already uses numerous custom content types, so I don't see it as a problem to introduce another.

**Comment by Jermolene** [comment]
> I would once again like to bump this.
> It doesn't have to be `text/x-tiddler-filter`, but as long as there's **some** way to differentiate filter tiddlers.

I still think it's a good idea, I think the only hurdle is deciding on the content type. I'm happy with `text/x-tiddler-filter`.

**Comment by pmario** [comment]
At https://github.com/Jermolene/TiddlyWiki5/blob/3f55f827a612742da481fb4597313aad9faf9591/boot/boot.js#L2379 we do register several TW specific types, but they are `application/...` 

So it would need to be as follows. 

´´´
$tw.utils.registerFileType("application/x-tiddler-filter","utf8",".filter",{deserializerType:"text/vnd.tiddlywiki"});
```

See the new extension: `.filter`

@Jermolene .. Is that right?

**Comment by flibbles** [comment]
I agree with @pmario that `application/x-tiddler-filter` is more apt than `text/x-tiddler-filter`. Filters are meant to be used, not looked at by end readers.

**Comment by flibbles** [comment]
I have not forgotten about this issue. Actually, I created the PR, but when I saw that the "Content Type" dropdown was now populated with yet another esoteric type, I came to a realization.

This addition would be feature creep.

There's a ton of good reasons not to add this.

1. End users are never going to use an `application/x-tiddler-filter` type, so it would just be another type cluttering up the "Content Type" dropdown that would confuse them.
2. When end-users edit filter tiddlers, it's almost always going to be through an `<$edit>` widget or something. They don't edit it directly, so it doesn't matter what the type is.
3. Developers probably aren't going to use it very often either. The only reason to do so is so that other plugins can properly recognize configuration and shadow tiddlers for their own esoteric reasons, and this is going to be a **very easy thing for them to forget to do**. And because of that, consistent use of the `application/x-tiddler-filter` type would be sporadic at best.
4. We'd be introducing support for a type that we could never take away because of the need for legacy support.

For all these reasons, I don't actually support this idea anymore, and I've been dragging my heels admitting this because I'm actually a little embarrassed. I was the one who spent so long endorsing this idea, so I've kinda wasted everyone's time. A far better solution as far as Relink is concerned is to develop its own way to recognize filter tiddlers that's not baked into the core.

If it's all right with everyone, I'd argue that it's best to close this proposal unmerged.

**Comment by pmario** [comment]
I think it's OK to close the issue. With v5.3.0 we start to get `\function()` pragma, which imo is very close to a "filter type". ... 

As you wrote, if you need dedicated tiddlers, that contain filters or tiddler lists the best way to go, if no rendering is needed, can be `text/plain`

We do have [filter config tiddlers](https://tiddlywiki.com/#%24%3A%2Fcore%2FFilters%2FOverriddenShadowTiddlers), which are used with the **Advanced Search -> Filter tab -> filter dropdown.** 

Those tiddlers have a `filter-field` and a tag: `$:/tags/Filter` .. It's one more possibility to work with filters. 

-----

For my bundler-plugin I also use `$:/tags/Bundle` and `type: text/plain` ... 

------

**Conclusion** ... I would be OK if it is closed. .. I also think, there is no need to be embarrassed ... It is a viable idea. .. But TW development may have made it obsolete.

**Comment by Jermolene** [comment]
> I have not forgotten about this issue. Actually, I created the PR, but when I saw that the "Content Type" dropdown was now populated with yet another esoteric type, I came to a realization.

Do we have to add that entry? As I remember it, the content type dropdown has explicitly defined entries that do not reflect the content types found in the wiki, so if we wanted to we could omit the new filter type from the list.

> There's a ton of good reasons not to add this.

Those are pretty good reasons, but there are also some valid reasons in the opposite direction. The notable one is that we would be able to apply a custom view template to display these filter tiddlers in a useful way: perhaps with a preview of the results of the filter, and maybe syntax highlighting if we put a bit of work in.

> For all these reasons, I don't actually support this idea anymore, and I've been dragging my heels admitting this because I'm actually a little embarrassed. I was the one who spent so long endorsing this idea, so I've kinda wasted everyone's time. A far better solution as far as Relink is concerned is to develop its own way to recognize filter tiddlers that's not baked into the core.

Its absolutely not a waste of time to have these discussions: it helps us to think things through, and ultimately leads to a better understanding of the task that we share.

> If it's all right with everyone, I'd argue that it's best to close this proposal unmerged.

I'm happy to close this, but I also still quite interested to see the PR...
